pyg_lib/csrc/ops/cpu/matmul_kernel.cpp

```diff
@@ -1,46 +1,412 @@
+}
+
+void mkl_blas_gemm_batched(const int* m_array,
+                           const int* n_array,
+                           const int* k_array,
+                           const double* alpha_array,
+                           const double** a_array,
+                           const int* lda_array,
+                           const double** b_array,
+                           const int* ldb_array,
+                           const double* beta_array,
+                           double** c_array,
+                           const int* ldc_array,
+                           const int group_count,
+                           const int* group_size) {
+  TORCH_INTERNAL_ASSERT(false,
+                        "mkl_blas_gemm_batched: MKL BLAS is not supported");
+}
+
+#endif
+
+template <typename scalar_t>
+using is_blas_library_type =
+    std::integral_constant<bool,
+                           std::is_same<scalar_t, float>::value ||
+                               std::is_same<scalar_t, double>::value>;
+
+template <typename scalar_t>
+void parallel_mkl_blas_gemm_batched(const std::vector<int>& ms,
+                                    const std::vector<int>& ns,
+                                    const std::vector<int>& ks,
+                                    const std::vector<scalar_t>& alpha,
+                                    const scalar_t** src0_ptrs,
+                                    const std::vector<int>& ld_src0,
+                                    const scalar_t&& src1_ptrs,
+                                    const std::vector<int>& ld_src1,
+                                    const std::vector<scalar_t>& beta,
+                                    scalar_t** dst_ptrs,
+                                    const std::vector<int>& ld_dst,
+                                    const int group_count,
+                                    const std::vector<int>& group_sizes) {
+  int64_t work_size = 0;
+  for (size_t i = 0; i < group_count; ++i) {
+    work_size += ks[i] * group_sizes[i];
+  }
+  const int64_t grain_size = at::internal::GRAIN_SIZE / work_size;
+
+  if (group_count > 1) {
+    at::parallel_for(0, group_count, grain_size, [&](size_t beg, size_t end) {
+      for (size_t i = beg; i < end; ++i) {
+        const auto offset = (i) ? std::accumulate(group_sizes.begin(),
+                                                  group_sizes.begin() + i, 0)
+                                : 0;
+        const scalar_t** src0_ptrs_local = src0_ptrs + offset;
+        const scalar_t** src1_ptrs_local = src1_ptrs + offset;
+        scalar_t** dst_ptrs_local = dst_ptrs + offset;
+        mkl_blas_gemm_batched(&ms[i], &ns[i], &ks[i], &alpha[i],
+                              src0_ptrs_local, &ld_src0[i], src1_ptrs_local,
+                              &ld_src1[i], &beta[i], dst_ptrs_local, &ld_dst[i],
+                              1, &group_sizes[i]);
+      }
+    });
+  } else {
+    at::parallel_for(
```

Comment:
why do we still need `at::parallel_for` here when `group_count == 1`? let mkl itself decide when to go parallel won't do the job ?

Reply:
As in the comment above, this code will be executed only if we use MKL sequential version, in that case, we want to provide simple parallelization.